src/http.rs

```diff
@@ -172,46 +171,21 @@ impl<B> fmt::Debug for HttpClient<B> {
     }
 }
 
-#[derive(Deserialize, Serialize, Clone, Debug)]
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq)]
 #[serde(deny_unknown_fields, rename_all = "snake_case", tag = "strategy")]
 pub enum Auth {
     Basic { user: String, password: String },
     Bearer { token: String },
 }
 
 impl Auth {
-    /// Get basic-auth credentials encoded in the url,
-    /// example: http://user:password@example.com/ .
-    /// Remove the credentials from the url if exist.
-    pub fn get_and_strip_basic_auth(url: &str) -> (String, Option<Self>) {
-        match Self::get_and_strip(url) {
-            Some((url, auth)) => (url, Some(auth)),
-            None => (url.to_owned(), None),
-        }
-    }
-
-    // We can use `?` with this return type.
-    fn get_and_strip(url: &str) -> Option<(String, Self)> {
-        let mut url = url::Url::parse(url).ok()?;
-
-        let user = url.username();
-        let scheme = url.scheme();
-        if !user.is_empty() && (scheme == "http" || scheme == "https") {
-            let user = percent_decode(user.as_bytes())
-                .decode_utf8_lossy()
-                .into_owned();
-
-            let password = url.password().unwrap_or("");
-            let password = percent_decode(password.as_bytes())
-                .decode_utf8_lossy()
-                .into_owned();
-
-            url.set_username("").ok()?;
-            url.set_password(None).ok()?;
-
-            Some((url.to_string(), Auth::Basic { user, password }))
+    /// If both is `Some`, return an error. Otherwise, choose one of them and store in `auth`.
+    pub fn merge_auth_config(auth: &mut Option<Auth>, auth1: &Option<Auth>) -> crate::Result<()> {
```

Comment:
I realize there is some overhead associated with this, but I'm wondering if this shouldn't just return a `crate::Result<Auth>` instead of mutating the first passed parameter. It could be tied to the lifetime to avoid cloning `Auth`:

```
pub fn merge_auth_config<'a>(auth: &'a Option<Auth>, auth1: &'a Option<Auth>) -> &'a crate::Result<Auth> {
```

Reply:
I personally would find that clearer as a user of this function.

Reply:
Thanks! I changed the signature to
```rust
fn merge_auth_config(auth: &Option<Auth>, auth1: &Option<Auth>) -> crate::Result<Option<Auth>>
```

Needed to return Option<Auth> because both can be `None`. I don't think returning a reference could work because of the way the function is used, we need an owned value:
```rust
authorization = Auth::merge_auth_config(&authorization, &uri.auth)?;
```

Reply:
I think he meant something like this, that could be used the same way without requiring any cloning:
```rust
fn merge_auth_config<'a>(auth1: &'a Option<Auth>, auth2: &'a Option<Auth>) -> crate::Result<&'a Option<Auth>>
```
Having said that, AFAICT this is only called during sink build, so the extra clone isn't huge.